types/react/index.d.ts

```diff
@@ -560,6 +560,147 @@ declare namespace React {
+    /**
+     * The signature is identical to `useEffect`, but it fires synchronously after all DOM mutations.
+     * Use this to read layout from the DOM and synchronously re-render. Updates scheduled inside
+     * `useLayoutEffect` will be flushed synchronously, before the browser has a chance to paint.
+     *
+     * Prefer the standard `useEffect` when possible to avoid blocking visual updates.
+     *
+     * If you’re migrating code from a class component, `useLayoutEffect` fires in the same phase as
+     * `componentDidMount` and `componentDidUpdate`.
+     *
+     * @version experimental
+     * @see https://reactjs.org/docs/hooks-reference.html#uselayouteffect
+     */
+    function useLayoutEffect(effect: EffectCallback, inputs?: InputIdentityList): void;
+    /**
+     * Accepts a function that contains imperative, possibly effectful code.
+     *
+     * @param effect Imperative function that can return a cleanup function
+     * @param inputs If present, effect will only activate if the values in the list change.
+     *
+     * @version experimental
+     * @see https://reactjs.org/docs/hooks-reference.html#useeffect
+     */
+    function useEffect(effect: EffectCallback, inputs?: InputIdentityList): void;
+    // NOTE: this does not accept strings, but this will have to be fixed by removing strings from type Ref<T>
+    /**
+     * `useImperativeMethods` customizes the instance value that is exposed to parent components when using
+     * `ref`. As always, imperative code using refs should be avoided in most cases.
+     *
+     * `useImperativeMethods` should be used with `React.forwardRef`.
+     *
+     * @version experimental
+     * @see https://reactjs.org/docs/hooks-reference.html#useimperativemethods
+     */
+    function useImperativeMethods<T>(ref: Ref<T>|undefined, init: () => T, inputs?: InputIdentityList): void;
+    // I made 'inputs' required here and in useMemo as there's no point to memoizing without the memoization key
+    // useCallback(X) is identical to just using X, useMemo(() => Y) is identical to just using Y.
+    /**
+     * `useCallback` will return a memoized version of the callback that only changes if one of the `inputs`
+     * has changed.
+     *
+     * @version experimental
+     * @see https://reactjs.org/docs/hooks-reference.html#usecallback
+     */
+    // TODO (TypeScript 3.0): <T extends (...args: never[]) => unknown>
+    function useCallback<T extends (...args: any[]) => any>(callback: T, inputs: InputIdentityList): T;
+    /**
+     * `useMemo` will only recompute the memoized value when one of the `inputs` has changed.
+     *
+     * Usage note: if calling `useMemo` with a referentially stable function, also give it as the input in
+     * the second argument.
+     *
+     * ```ts
+     * function expensive () { ... }
+     *
+     * function Component () {
+     *   const expensiveResult = useMemo(expensive, [expensive])
+     *   return ...
+     * }
+     * ```
```

Comment:
This looks weird in the tooltip:

![image](https://user-images.githubusercontent.com/73085/47829298-14638980-ddca-11e8-82be-602734e0425c.png)

but formatting it in a way that gets vscode to format it correctly gives me a very angry tslint.